File: week1/Ariling/prgs-단어변환/main.py
def solution(begin, target, words):
    answer = int(1e9)
    checked_arr = [False] * len(words)
    # 알파벳 바꾸기 가능 여부 함수
    def diff_check(current, compare):
        diff = 0
        for i in range(len(current)):
            if current[i] != compare[i]:
                diff += 1
        if diff > 1:
            return False
        return True
    # DFS활용(재귀방식)
    def convert_word(c, t, check, try_num):
        nonlocal answer
        if t not in words:
            answer = 0
            return
        if try_num > answer:
            return
        if c == t and try_num < answer:
            answer = try_num
            return
        for i in range(len(words)):
            if not check[i] and diff_check(c, words[i]):
                check[i] = True
                convert_word(words[i], t, check, try_num + 1)
                check[i] = False
    convert_word(begin, target, checked_arr, 0)
    return answer

File: week1/Ariling/prgs-단어변환/main.js
function solution(begin, target, words) {
  var answer = 0;
  var min_len = words.length + 1;
  function compare_possible(target, compare) {
    var diff = 0;
    for (let x = 0; x < target.length; x++) {
      if (target[x] !== compare[x]) {
        diff += 1;
      }
    }
    if (diff === 1) {
      return true;
    } else {
      return false;
    }
  }
  var check_arr = Array(words.length).fill(false);
  function find_word(origin, target, check, count) {
    if (!words.includes(target)) {
      return;
    }
    if (count >= min_len) {
      return;
    }
    if (origin === target && count < min_len) {
      answer = count;
      min_len = count;
      return;
    }
    for (let x = 0; x < words.length; x++) {
      if (compare_possible(origin, words[x]) && !check[x]) {
        check[x] = true;
        find_word(words[x], target, check, count + 1);
        check[x] = false;
      }
    }
  }
  find_word(begin, target, check_arr, 0);
  return answer;
}
